agent/agent.go

```diff
@@ -588,79 +578,113 @@ func (r *agent) getStatusFrom(ctx context.Context, member serf.Member, respc cha
 	}
 }
 
-// statusResponse describes a status response from a background process that obtains
-// health status on the specified serf node.
-type statusResponse struct {
-	*pb.NodeStatus
-	member serf.Member
-	err    error
-}
+// timelineUpdateLoop periodically updates the status timeline by comparing
+// changes in local status and merging a timeline from another member.
+func (r *agent) timelineUpdateLoop() {
+	ticker := r.timelineClock.NewTicker(statusUpdateTimeout)
+	defer ticker.Stop()
 
-// recentStatus returns the last known cluster status.
-func (r *agent) recentStatus() (status *pb.SystemStatus, err error) {
-	status, err = r.cache.RecentStatus()
-	if err == nil && status == nil {
-		status = pb.EmptyStatus()
+	for range ticker.Chan() {
+		if r.isDone() {
+			return
+		}
+
+		if err := r.updateTimeline(); err != nil {
+			log.WithError(err).Warnf("Error updating timeline.")
+		}
 	}
-	return status, trace.Wrap(err)
 }
 
-// recentLocalStatus returns the last known local node status.
-func (r *agent) recentLocalStatus() *pb.NodeStatus {
-	r.mu.Lock()
-	defer r.mu.Unlock()
-	return r.localStatus
-}
+// updateTimeline updates the timeline with any new status changes.
+func (r *agent) updateTimeline() error {
+	ctx, cancel := context.WithTimeout(context.Background(), r.statusQueryReplyTimeout)
+	defer cancel()
+
+	// Update local status changes
+	if status := r.LocalStatus(); status != nil {
+		if err := r.Timeline.RecordStatus(ctx, status); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	// Collect timeline from another cluster member
+	timeline, err := r.collectTimeline(ctx)
```

Comment:
Didn't we want to have all nodes send their timeline events to the masters so master nodes will have a consistent view of the entire timeline?